skbio/sequence/_substitution.py

```diff
@@ -0,0 +1,518 @@
+        name : str
+            Name of the substitution matrix.
+
+        Returns
+        -------
+        SubstitutionMatrix
+            Named substitution matrix.
+
+        Raises
+        ------
+        ValueError
+            If named substitution matrix does not exist.
+
+        See Also
+        --------
+        get_names
+
+        Notes
+        -----
+        Available substitution matrix names can be obtained by ``get_names``.
+        Currently, the following names are supported:
+
+        - `NUC.4.4` (a.k.a. DNAfull): A nucleotide substitution matrix covering
+          all definite and degenerate nucleotides.
+
+        - Point Accepted Mutation (PAM) [1]_: A set of amino acid substitution
+          matrices, including `PAM30`, `PAM70` and `PAM250`.
+
+        - BLOcks SUbstitution Matrix (BLOSUM) [2]_: A set of amino acid
+          substitution matrices, including `BLOSUM45`, `BLOSUM50`, `BLOSUM62`,
+          `BLOSUM80` and `BLOSUM90`.
+
+        References
+        ----------
+        .. [1] Dayhoff, M., Schwartz, R., & Orcutt, B. (1978). A model of
+           evolutionary change in proteins. Atlas of protein sequence and
+           structure, 5, 345-352.
+        .. [2] Henikoff, S., & Henikoff, J. G. (1992). Amino acid substitution
+           matrices from protein blocks. Proceedings of the National Academy of
+           Sciences, 89(22), 10915-10919.
+
+        Examples
+        --------
+        >>> from skbio import SubstitutionMatrix
+        >>> mat = SubstitutionMatrix.by_name('BLOSUM62')
+        >>> len(mat.alphabet)
+        24
+        >>> mat['M', 'K']
+        -1.0
+
+        """
+        try:
+            return named_substitution_matrices[name]
+        except KeyError:
+            raise ValueError(f'Substitution matrix "{name}" does not exist.')
+
+    @classonlymethod
+    @experimental(as_of='0.5.10')
+    def get_names(cls):
+        """List names of pre-defined substitution matrices.
+
+        Returns
+        -------
+        list of str
+            Names of pre-defined substitution matrices.
+
+        See Also
+        --------
+        by_name
+        """
+        return list(named_substitution_matrices.keys())
+
+
+def _matrix_to_vector(mat):
+    """Convert a square matrix into a flattened vector consisting of the upper
+    triangle and the diagonal.
+    """
+    return mat[np.triu_indices(len(mat))]
+
+
+def _vector_to_matrix(vec):
+    """Convert a vector consisting of the upper triangle and the diagonal of a
+    matrix to square form.
+    """
+    n = int((np.sqrt(1 + 8 * len(vec)) - 1) / 2)
+    mat = np.zeros((n, n))
+    mat[np.triu_indices(n)] = vec
+    return mat + np.triu(mat, k=1).T
+
+
+# Defined according to the matrices hosted at the NCBI FTP server:
+# https://ftp.ncbi.nlm.nih.gov/blast/matrices/
+named_substitution_matrices = {
+
+    # NUC.4.4, a.k.a. DNAfull
+    'NUC.4.4': SubstitutionMatrix(
+        'ATGCSWRYKMBVHDN', _vector_to_matrix(np.array([
```

Comment:
I though a bit but am not sure what's the best strategy to store the matrix data. One way is to keep them as external files (like [this one](https://ftp.ncbi.nlm.nih.gov/blast/matrices/BLOSUM62)). If we keep the data in the Python script, we cannot make the values look nice anyway because each row will exceed 80 characters and we have to break them into multiple lines.

Reply:
`# noqa` can always be added to the end of the lines. But I really like the idea of just storing or downloading the files, as we can assert they are identical to what's provided by ncbi

Reply:
At this moment, skbio cannot directly parse the matrix files correctly. I added this functionality (see #1912 ) but it is not perfect yet. How about we keep the current solution (in Python code) for now, and have a separate discussion on how to store external data in skbio? I envision that such a mechanism will be highly useful (like `load_iris` in some other Python/R packages).

Reply:
Makes sense. And also like `biom.example_table`... having easy to get data is awesome

Reply:
@wasade I submitted an issue in SciPy: https://github.com/scipy/scipy/issues/19875